dask_expr/io/parquet.py

```diff
@@ -1399,3 +1594,183 @@ def _normalize_and_strip_protocol(path):
+    file_level_stats = ["num_rows", "num_row_groups", "serialized_size"]
+    rg_stats = [
+        "num_rows",
+        "total_byte_size",
+        "sorting_columns",
+    ]
+    col_meta = [
+        "num_values",
+        "total_compressed_size",
+        "total_uncompressed_size",
+        "path_in_schema",
+    ]
+    col_stats = [
+        "min",
+        "max",
+        "null_count",
+        "num_values",
+        "distinct_count",
+    ]
+
+    out = {}
+    for name in file_level_stats:
+        out[name] = original[name]
+    out["row_groups"] = rgs = []
+    for rg in original["row_groups"]:
+        rg_out = {}
+        rgs.append(rg_out)
+        for name in rg_stats:
+            rg_out[name] = rg[name]
+        rg_out["columns"] = []
+        for col in rg["columns"]:
+            col_out = {}
+            rg_out["columns"].append(col_out)
+            for name in col_meta:
+                col_out[name] = col[name]
+            col_out["statistics"] = {}
+            for name in col_stats:
+                col_out["statistics"][name] = col["statistics"][name]
+
+    return out
+
+
+def _agg_dicts(dicts, agg_funcs):
+    result = {}
+    for d in dicts:
+        for k, v in d.items():
+            if k not in result:
+                result[k] = [v]
+            else:
+                result[k].append(v)
+    result2 = {}
+    for k, v in result.items():
+        agg = agg_funcs.get(k)
+        if agg:
+            result2[k] = agg(v)
+    return result2
+
+
+def _aggregate_columns(cols, agg_cols):
+    combine = []
+    i = 0
+    while True:
+        inner = []
+        combine.append(inner)
+        try:
+            for col in cols:
+                inner.append(col[i])
+        except IndexError:
+            combine.pop()
+            break
+        i += 1
+    return [_agg_dicts(c, agg_cols) for c in combine]
+
+
+def _aggregate_statistics_to_file(stats):
+    """Aggregate RG information to file level."""
+
+    agg_stats = {
+        "min": min,
+        "max": max,
+    }
+    agg_cols = {
+        "total_compressed_size": sum,
+        "total_uncompressed_size": sum,
+        "statistics": partial(_agg_dicts, agg_funcs=agg_stats),
+        "path_in_schema": lambda x: set(x).pop(),
+    }
+    agg_func = {
+        "num_rows": sum,
+        "total_byte_size": sum,
+        "columns": partial(_aggregate_columns, agg_cols=agg_cols),
+    }
+    aggregated_stats = []
+    for file_stat in stats:
+        file_stat = file_stat.copy()
+        aggregated_stats.append(file_stat)
+
+        file_stat.update(_agg_dicts(file_stat.pop("row_groups"), agg_func))
+    return aggregated_stats
+
+
+@dask.delayed
+def _gather_statistics(frags):
+    @dask.delayed
+    def _collect_statistics(token_fragment):
+        return token_fragment[0], _extract_stats(token_fragment[1].metadata.to_dict())
+
+    return dask.compute(
+        list(_collect_statistics(frag) for frag in frags), scheduler="threading"
```

Comment:
Just double checking my understanding here: This operates on the stats cache that is stored locally, so running it locally is what we want, correct?

Reply:
`_gather_statistics` (the outer function) is being dispatched to a cluster if one is available while the inner function `_collect_statistics` is executed on that cluster inside of a worker and is being parallelized using a threadpool.

typing this now makes me wonder how well dask handles nested concurrency that is accessing the same threadpool. I could see this deadlock if it's using the same global pool. I'll double check.

Reply:
nope, everything good with the threading thing. We may spawn many threads but that's fine. Dask is handling this properly and ensures that the nested call get it's own threadpool